### PR TITLE
Restrict workshop material order user requirement to only when user_id changes

### PR DESCRIPTION
**Pull request by aoby**
to allow users who are deleted after ordering

Error in the wild: https://app.honeybadger.io/projects/45435/faults/33629810#notice-summary
```Validation failed: User is required (ActiveRecord::RecordInvalid)```

### CODE REVIEW

**Comment by aoby** [comment]
For more context, in this case, the user must have existed to create the order, so this happens when a user successfully creates an order and then later (soft) deletes their account. At that point we should still be able to track the order and log status.

File: dashboard/app/models/pd/workshop_material_order.rb [hunk 1]
```diff
@@ -50,7 +50,7 @@ class WorkshopMaterialOrder < ActiveRecord::Base
     has_one :workshop, class_name: 'Pd::Workshop', through: :enrollment, foreign_key: :pd_workshop_id
 
     validates :enrollment, presence: true, uniqueness: true
-    validates :user, presence: true, uniqueness: true
+    validates :user, presence: true, uniqueness: true, if: -> {user_id_changed?}
```
Comment (ashercodeorg) [inline dashboard/app/models/pd/workshop_material_order.rb]: Unfortunately, I think `user_id_changed?` is false for a newly created `Pd::WorkshopMaterialOrder` without a user, making the validation mostly useless? Note that I'm lacking context here.

File: dashboard/test/models/pd/workshop_material_order_test.rb [hunk 2]
```diff
@@ -301,4 +301,18 @@ class Pd::WorkshopMaterialOrderTest < ActiveSupport::TestCase
     order.refresh
     assert_nil order.tracking_id
   end
+
+  test 'orders are still valid if the user is deleted' do
```
Comment (ashercodeorg) [inline dashboard/test/models/pd/workshop_material_order_test.rb]: No changes expected, but grumble, grumble. The `assert_equal` is not part of the test asserts, it is asserting a precondition. And this is two tests, one confirming that the order is valid with a deleted user, another confirming that the order is invalid after a user change.

**Comment by ashercodeorg** [comment]
Th approach I'm fine with, I don't think the implementation's side effects are desired.

File: dashboard/test/models/pd/workshop_material_order_test.rb (hunk 2 above)
Comment (aoby) [reply dashboard/test/models/pd/workshop_material_order_test.rb]: I'll remove the initial assert. As far as 2 tests, would you prefer to separate them? That seems extra cumbersome, given that we don't (yet?) support describe/it syntax with nested setups.

File: dashboard/app/models/pd/workshop_material_order.rb (hunk 1 above)
Comment (aoby) [reply dashboard/app/models/pd/workshop_material_order.rb]: I don't think so. Note [this](https://github.com/code-dot-org/code-dot-org/blob/885d2d43cfca009506590324f56a06318f89cf88/dashboard/test/models/pd/workshop_material_order_test.rb#L53-L53) `User is required` error is still occurring on a new model. I can add an explicit test for the nil user to be sure.

File: dashboard/test/models/pd/workshop_material_order_test.rb (hunk 2 above)
Comment (ashercodeorg) [reply dashboard/test/models/pd/workshop_material_order_test.rb]: As I said, no changes expected. But tests should test one and only one thing, IMO.

File: dashboard/app/models/pd/workshop_material_order.rb (hunk 1 above)
Comment (ashercodeorg) [reply dashboard/app/models/pd/workshop_material_order.rb]: Any idea why I'm seeing this locally? (using some recent version of staging)
```
cosmiccrisp:code-dot-org asher$ ./bin/dashboard-console 
Running via Spring preloader in process 76437
Loading development environment (Rails 5.0.1)
irb: warn: can't alias context from irb_context.
irb(main):001:0> x = Pd::WorkshopMaterialOrder.new
=> #<Pd::WorkshopMaterialOrder id: nil, created_at: nil, updated_at: nil, pd_enrollment_id: nil, user_id: nil, school_or_company: nil, street: nil, apartment_or_suite: nil, city: nil, state: nil, zip_code: nil, phone_number: nil, order_attempted_at: nil, ordered_at: nil, order_response: nil, order_error: nil, order_id: nil, order_status: nil, order_status_last_checked_at: nil, order_status_changed_at: nil, tracking_id: nil, tracking_url: nil>
irb(main):002:0> x.user_id_changed?
=> false
```

Comment (aoby) [reply dashboard/app/models/pd/workshop_material_order.rb]: Actually you're right. That test indeed is failing :(

Comment (aoby) [reply dashboard/app/models/pd/workshop_material_order.rb]: Yeah, you're right. `user_id_changed?` is false on a new model. However this works: `new_record? || user_id_changed?`. Updating...

Comment (ashercodeorg) [reply dashboard/app/models/pd/workshop_material_order.rb]: LGTM with that change. Not sure I remembered `new_record?`. Too many magical Rails methods...

**Pull request merged**